### PR TITLE
Source Code Syntax Highlighter 1.0

### PR DESCRIPTION
**Pull request by karupayun**
This PR upgrades the C Syntax Highlighter adding C++ rules. More
specific:

 - Upgrades number and keyword regex.

 - Modifies function regex to only include function definitions
 (as discussed)

 - Introduces class members and add examples (track_, m_track, ..)

 - Introduces comma regex (to paint ',' and ';'), and also modifies
 capitalized words (to not paint '*' and '&' in variables).

 - Update few colors following CLion color conventions and uses a new
 one to paint class' names and types like std::vector.

Screenshots:
file
tests

To test it:
 - Run CodeViewerDemo. You will see there some examples and a real file.

Last step in http://b/177651839.

### CODE REVIEW

File: src/CodeViewerDemo/CodeExamples.h
Comment (florian-kuebler) [inline src/CodeViewerDemo/CodeExamples.h:29]: If I am not wrong, the code style here would be `m_PascalCase`. Not sure if that matters.

File: src/CodeViewerDemo/CodeExamples.h
Comment (florian-kuebler) [inline src/CodeViewerDemo/CodeExamples.h:35]: should this end with a `;`? Similar above.

File: src/CodeViewerDemo/CodeExamples.h
Comment (florian-kuebler) [inline src/CodeViewerDemo/CodeExamples.h:32]: Should you add a closing `}`?

File: src/SyntaxHighlighter/Cpp.cpp
Comment (florian-kuebler) [inline src/SyntaxHighlighter/Cpp.cpp:50]: so you don't allow `_` in function names? Remember Google Style Guide as `snake_case()`/`set_snake_case(T t)` for simple accessor methods. If that is not easily doable, also feel free to ignore.

**Review by florian-kuebler — APPROVED** [review]
That's pretty nice! LGTM!

File: src/CodeViewerDemo/CodeExamples.h
Comment (karupayun) [reply src/CodeViewerDemo/CodeExamples.h:29]: Doesn't matter, but I will change it.

File: src/CodeViewerDemo/CodeExamples.h
Comment (karupayun) [reply src/CodeViewerDemo/CodeExamples.h:35]: I will polish a little bit CodeExamples to make code which have some sense.

File: src/SyntaxHighlighter/Cpp.cpp
Comment (beckerhe) [reply src/SyntaxHighlighter/Cpp.cpp:50]: `\w` includes a-z, A-Z, 0-9 ,and "_".

Comment (beckerhe) [reply src/SyntaxHighlighter/Cpp.cpp:50]: The last `\w*` should probably be `\w+`, no? And the one before `::` as well, isn't it?

File: src/SyntaxHighlighter/Cpp.cpp
Comment (beckerhe) [inline src/SyntaxHighlighter/Cpp.cpp:64]: Why is there a differentiation between upper and lower case?

File: src/SyntaxHighlighter/Cpp.cpp
Comment (beckerhe) [inline src/SyntaxHighlighter/Cpp.cpp:40]: Nice, we already support `concept` and `requires`. You could also add `import` and `export`.

File: src/SyntaxHighlighter/Cpp.cpp
Comment (beckerhe) [inline src/SyntaxHighlighter/Cpp.cpp:125]: Nit: comment. comment.

**Review by beckerhe — APPROVED** [review]
Very nice Pablo. Thank you a lot for that!

File: src/SyntaxHighlighter/Cpp.cpp
Comment (karupayun) [reply src/SyntaxHighlighter/Cpp.cpp:50]: \w includes '_' :)
You can see 2 examples there in the screenshots with functions having '_'.

I'm differentiating them using a horrible heuristic here. I'm asking for a character before the function(

In that way:
"int function();'' has the 't' of int
and "lets_execute_function()" not.

"(?<=\\w)\\s+" is doing that.
?<= means looking behind without painting it
and \\s+ is the spaces between that character and the function itself.

BTW, thanks for your comment, I've just realized that I have to polish this a little bit to avoid having number as a first character here. :)

File: src/SyntaxHighlighter/Cpp.cpp
Comment (karupayun) [reply src/SyntaxHighlighter/Cpp.cpp:125]: Ohh, sorry! :(

Comment (beckerhe) [reply src/SyntaxHighlighter/Cpp.cpp:125]: No worries. :slightly_smiling_face:

File: src/SyntaxHighlighter/Cpp.cpp
Comment (karupayun) [reply src/SyntaxHighlighter/Cpp.cpp:64]: Nice catch! I started making different things and after changed it and never review after that.

File: src/SyntaxHighlighter/Cpp.cpp
Comment (karupayun) [reply src/SyntaxHighlighter/Cpp.cpp:50]: Yes, thanks Henning

File: src/SyntaxHighlighter/Cpp.cpp
Comment (karupayun) [reply src/SyntaxHighlighter/Cpp.cpp:64]: I should rewrite this to include lower and uppercase together. Thanks for pointing out. It's something that I used to need, but not anymore, I think.

It isn't a difference between lower and upper, it's about the left and right. The first part is catching the left part of the " : : " and the second the right part which shouldn't be a function (because we are painting them yellow).

File: src/SyntaxHighlighter/Cpp.cpp
Comment (dpallotti) [inline src/SyntaxHighlighter/Cpp.cpp:14]: I would suggest a quick comment on where these colors are taken from (CLion Darcula?), for documentation purposes and in case one wants to add more.

Comment (karupayun) [reply src/SyntaxHighlighter/Cpp.cpp:15]: better?

File: src/SyntaxHighlighter/Cpp.cpp
Comment (karupayun) [reply src/SyntaxHighlighter/Cpp.cpp]: Changed: "(?<=\\w)\\s+([A-Za-z_]\\w*::)*[A-Za-z_]\\w*(?=\\()"

Instead of **_\w+_**, I used **_[A-Za-z_]\w*_**, because it shouldn't start with a number

File: src/SyntaxHighlighter/Cpp.cpp
Comment (karupayun) [reply src/SyntaxHighlighter/Cpp.cpp]: Divided into two:
kNamespaceRegex match the first part before "::" and kNamespaceVariableRegex match the right part. Turn out to be simpler after that division and cleaner.

File: src/SyntaxHighlighter/Cpp.cpp
Comment (dpallotti) [reply src/SyntaxHighlighter/Cpp.cpp:15]: Thanks (don't wait for my approval)

**Review by florian-kuebler — COMMENTED** [review]
Thanks for the update. LGTM!

**Pull request merged**
